File: kernel/internal/session/account.go
package session

import (
	"errors"
	"sync"
)

// ResourceQuota 是单个 agent 的资源预算。
type ResourceQuota struct {
	MaxSteps  int
	MaxTokens int
	// MaxDuration / MaxToolCalls 留后续
}

// Usage 是已消耗的资源。
type Usage struct {
	Steps  int
	Tokens int
}

// Account 累计资源消耗，超限拒绝。并发安全。
type Account struct {
	mu    sync.Mutex
	quota ResourceQuota
	used  Usage
}

func NewAccount(q ResourceQuota) *Account {
	return &Account{quota: q}
}

// Charge 扣减资源。任一项超限返回 ErrQuotaExceeded（但已累加，调用方应终止）。
func (a *Account) Charge(u Usage) error {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.used.Steps += u.Steps
	a.used.Tokens += u.Tokens
	if a.quota.MaxSteps > 0 && a.used.Steps > a.quota.MaxSteps {
		return ErrQuotaExceeded
	}
	if a.quota.MaxTokens > 0 && a.used.Tokens > a.quota.MaxTokens {
		return ErrQuotaExceeded
	}
	return nil
}

// Used 返回当前累计消耗（用于审计/展示）。
func (a *Account) Used() Usage {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.used
}

var ErrQuotaExceeded = errors.New("quota exceeded")

File: cp/session/account.py
import asyncio
from dataclasses import dataclass


class QuotaExceeded(Exception):
    pass


@dataclass
class ResourceQuota:
    max_steps: int = 0
    max_tokens: int = 0


@dataclass
class Usage:
    steps: int = 0
    tokens: int = 0


class Account:
    """累计资源消耗,超限拒绝。asyncio.Lock 保护(协程安全)。"""

    def __init__(self, quota: ResourceQuota):
        self._mu = asyncio.Lock()
        self.quota = quota
        self._used = Usage()

    async def charge(self, u: Usage) -> None:
        """扣减资源。任一项超限抛 QuotaExceeded(但已累加,调用方应终止)。"""
        async with self._mu:
            self._used.steps += u.steps
            self._used.tokens += u.tokens
            if self.quota.max_steps > 0 and self._used.steps > self.quota.max_steps:
                raise QuotaExceeded()
            if self.quota.max_tokens > 0 and self._used.tokens > self.quota.max_tokens:
                raise QuotaExceeded()

    async def used(self) -> Usage:
        async with self._mu:
            return Usage(steps=self._used.steps, tokens=self._used.tokens)
